### PR TITLE
Django 1.8

### PR DESCRIPTION
**Pull request by nickpell**
opening for tests
- [x] use django-crispy-forms 1.6.0

### CODE REVIEW

**Comment by benrudolph** [comment]
is this possible without switching our test runner?

File: corehq/apps/style/forms/widgets.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 from django import forms
 from django.forms.fields import MultiValueField, CharField
-from django.forms.utils import flatatt
-from django.forms.widgets import CheckboxInput, HiddenInput, Input, RadioSelect, RadioFieldRenderer, RadioInput, TextInput, MultiWidget
+from django.forms.util import flatatt
+from django.forms.widgets import CheckboxInput, HiddenInput, Input, RadioSelect, RadioFieldRenderer, RadioChoiceInput, TextInput, MultiWidget
```
Comment (dimagimon) [inline corehq/apps/style/forms/widgets.py]: F401 'HiddenInput' imported but unused

Comment (dimagimon) [inline corehq/apps/style/forms/widgets.py]: E501 line too long (141 > 115 characters)

File: corehq/apps/style/forms/widgets.py [hunk 2]
```diff
@@ -30,7 +30,7 @@ def render(self, name, value, attrs=None):
         return mark_safe(u'<label class="checkbox"><input%s /> %s</label>' %
                          (flatatt(final_attrs), self.inline_label))
 
-class BootstrapRadioInput(RadioInput):
+class BootstrapRadioInput(RadioChoiceInput):
```
Comment (dimagimon) [inline corehq/apps/style/forms/widgets.py]: E302 expected 2 blank lines, found 1

**Comment by nickpell** [comment]
@benrudolph I don't have a definitive answer to that at the moment.  In the meantime I'd like to take care of other blockers (like 5a315bcef5da6a1a445445c2f23e951521915697) so we're closer to being able to upgrade

**Comment by benrudolph** [comment]
legit. can't wait for this

**Comment by esoergel** [comment]
:star2:

**Comment by nickpell** [comment]
latest test failure due to https://groups.google.com/forum/#!msg/django-users/zCi-Nhm-tmY/9ThHwuHAxbsJ

**Comment by nickpell** [comment]
latest error:

```
Traceback (most recent call last):
  File "manage.py", line 96, in <module>
    execute_from_command_line(sys.argv)
  File "/vendor/lib/python2.7/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/vendor/lib/python2.7/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/vendor/lib/python2.7/site-packages/django/core/management/commands/test.py", line 30, in run_from_argv
    super(Command, self).run_from_argv(argv)
  File "/vendor/lib/python2.7/site-packages/django/core/management/base.py", line 382, in run_from_argv
    parser = self.create_parser(argv[0], argv[1])
  File "/vendor/lib/python2.7/site-packages/django/core/management/base.py", line 355, in create_parser
    self.add_arguments(parser)
  File "/vendor/lib/python2.7/site-packages/django/core/management/commands/test.py", line 62, in add_arguments
    test_runner_class.add_arguments(parser)
  File "/vendor/lib/python2.7/site-packages/django_nose/runner.py", line 200, in add_arguments
    config.plugins.addPlugins(list(_get_plugins_from_settings()))
  File "/vendor/lib/python2.7/site-packages/django_nose/runner.py", line 107, in _get_plugins_from_settings
    'Error importing Nose plugin module %s: "%s"' % (p_mod, e))
ImproperlyConfigured: Error importing Nose plugin module corehq.tests.nose: "No module named creation"
```

**Comment by millerdev** [comment]
@nickpell I'll take a look next week

**Comment by nickpell** [comment]
```
Traceback (most recent call last):
  File "manage.py", line 96, in <module>
    execute_from_command_line(sys.argv)
  File "/vendor/lib/python2.7/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/vendor/lib/python2.7/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/vendor/lib/python2.7/site-packages/django/core/management/commands/test.py", line 30, in run_from_argv
    super(Command, self).run_from_argv(argv)
  File "/vendor/lib/python2.7/site-packages/django/core/management/base.py", line 382, in run_from_argv
    parser = self.create_parser(argv[0], argv[1])
  File "/vendor/lib/python2.7/site-packages/django/core/management/base.py", line 355, in create_parser
    self.add_arguments(parser)
  File "/vendor/lib/python2.7/site-packages/django/core/management/commands/test.py", line 62, in add_arguments
    test_runner_class.add_arguments(parser)
  File "/vendor/lib/python2.7/site-packages/django_nose/runner.py", line 200, in add_arguments
    config.plugins.addPlugins(list(_get_plugins_from_settings()))
  File "/vendor/lib/python2.7/site-packages/django_nose/runner.py", line 107, in _get_plugins_from_settings
    'Error importing Nose plugin module %s: "%s"' % (p_mod, e))
ImproperlyConfigured: Error importing Nose plugin module corehq.tests.nose: "No module named simple"
```

**Comment by nickpell** [comment]
above error caused by the removal of ```django.test.simple```

**Comment by nickpell** [comment]
@millerdev thanks, I figured out the first error.  Now getting a different error originally from the hq test runner

File: corehq/tests/nose.py [hunk 3]
```diff
@@ -26,7 +26,11 @@
 from nose.plugins import Plugin
 from django.apps import AppConfig
 from django.conf import settings
-from django.db.backends.creation import TEST_DATABASE_PREFIX
+try:
+    from django.db.backends.base.creation import TEST_DATABASE_PREFIX
+except ImportError:
+    # TODO - remove when django >= 1.8
+    from django.db.backends.creation import TEST_DATABASE_PREFIX
```
Comment (millerdev) [inline corehq/tests/nose.py]: do we also need this in [settingshelper.py](https://github.com/dimagi/commcare-hq/blob/a57fdc1ec617e75ab1d66114a41241ad74be5efe/settingshelper.py#L7)?

**Comment by millerdev** [comment]
re: `ImproperlyConfigured: Error importing Nose plugin module corehq.tests.nose: "No module named simple"`

Looks like the culprit might be [this line](https://github.com/dimagi/commcare-hq/blob/a57fdc1ec617e75ab1d66114a41241ad74be5efe/corehq/tests/optimizer.py#L9). @czue: looks like the optimizer may need to be overhauled for Django 1.8

**Comment by millerdev** [comment]
Maybe the fix is to switch from using `build_test` to [`build_suite`](https://docs.djangoproject.com/en/1.9/topics/testing/advanced/#django.test.runner.DiscoverRunner.build_suite)?

**Comment by millerdev** [comment]
Might need to merge #11713 and then master into this branch for 3bad9ae to work here.

**Comment by millerdev** [comment]
tests failed (all three nodes failed with the same error):
```
======================================================================
ERROR: test_report_module (corehq.apps.app_manager.tests.test_suite.SuiteTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/apps/app_manager/tests/test_suite.py", line 659, in test_report_module
    from corehq.apps.userreports.tests.utils import get_sample_report_config
  File "/mnt/corehq/apps/userreports/tests/__init__.py", line 2, in <module>
    from .test_columns import *
  File "/mnt/corehq/apps/userreports/tests/test_columns.py", line 26, in <module>
    from casexml.apps.case.tests.util import delete_all_cases
  File "corehq/ex-submodules/casexml/apps/case/tests/__init__.py", line 5, in <module>
    from casexml.apps.case.tests.util import delete_all_cases, delete_all_xforms
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 7, in <module>
    from corehq.form_processor.tests.utils import FormProcessorTestUtils
  File "/mnt/corehq/form_processor/tests/__init__.py", line 4, in <module>
    from .test_form_dbaccessor import *
  File "/mnt/corehq/form_processor/tests/test_form_dbaccessor.py", line 21, in <module>
    from crispy_forms.tests.utils import override_settings
  File "submodules/bootstrap3_crispy/crispy_forms/tests/__init__.py", line 1, in <module>
    from .test_tags import *
  File "submodules/bootstrap3_crispy/crispy_forms/tests/test_tags.py", line 8, in <module>
    from .forms import TestForm
  File "submodules/bootstrap3_crispy/crispy_forms/tests/forms.py", line 89, in <module>
    class TestForm4(forms.ModelForm):
  File "/vendor/lib/python2.7/site-packages/django/forms/models.py", line 274, in __new__
    "needs updating." % name
ImproperlyConfigured: Creating a ModelForm without either the 'fields' attribute or the 'exclude' attribute is prohibited; form TestForm4 needs updating.
```

File: requirements/requirements.txt [hunk 4]
```diff
@@ -7,7 +7,7 @@ jsonobject-couchdbkit==0.7.4.0
 celery==3.1.18
 # required by django-digest
 decorator==4.0.4
-django==1.7.11
+django==1.8.11
```
Comment (emord) [inline requirements/requirements.txt]: I think this can be bumped to 1.8.13

**Comment by nickpell** [comment]
Getting a lot of test errors like:
```
======================================================================
ERROR: testBasicFunctionality (corehq.apps.reports.tests.test_cache.ReportCacheTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/apps/reports/tests/test_cache.py", line 51, in setUp
    self.test_domain = create_domain(self.domain)
  File "/mnt/corehq/apps/domain/shortcuts.py", line 9, in create_domain
    secure_submissions=False)
  File "/mnt/corehq/apps/domain/models.py", line 556, in get_or_create_with_name
    new_domain.save(**get_safe_write_kwargs())
  File "/mnt/corehq/apps/domain/models.py", line 617, in save
    set_migration_complete(self.name)
  File "/mnt/corehq/apps/tzmigration/api.py", line 39, in set_migration_complete
    progress, _ = TimezoneMigrationProgress.objects.get_or_create(pk=domain)
  File "/vendor/lib/python2.7/site-packages/django/db/models/manager.py", line 127, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/db/models/query.py", line 405, in get_or_create
    return self.get(**lookup), False
  File "/vendor/lib/python2.7/site-packages/django/db/models/query.py", line 328, in get
    num = len(clone)
  File "/vendor/lib/python2.7/site-packages/django/db/models/query.py", line 144, in __len__
    self._fetch_all()
  File "/vendor/lib/python2.7/site-packages/django/db/models/query.py", line 965, in _fetch_all
    self._result_cache = list(self.iterator())
  File "/vendor/lib/python2.7/site-packages/django/db/models/query.py", line 238, in iterator
    results = compiler.execute_sql()
  File "/vendor/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 838, in execute_sql
    cursor = self.connection.cursor()
  File "/vendor/lib/python2.7/site-packages/django/db/backends/base/base.py", line 164, in cursor
    cursor = self.make_cursor(self._cursor())
  File "/vendor/lib/python2.7/site-packages/django/db/backends/base/base.py", line 137, in _cursor
    return self.create_cursor()
  File "/vendor/lib/python2.7/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/vendor/lib/python2.7/site-packages/django/db/backends/base/base.py", line 137, in _cursor
    return self.create_cursor()
  File "/vendor/lib/python2.7/site-packages/django/db/backends/postgresql_psycopg2/base.py", line 212, in create_cursor
    cursor = self.connection.cursor()
InterfaceError: connection already closed
```
Other people have reported the same problem (http://osdir.com/ml/general/2015-07/msg16599.html) but I couldn't find a solution that isn't already in place here.

**Comment by millerdev** [comment]
does that test (or the one before it) call a managment command?

from the [release notes](https://docs.djangoproject.com/en/1.9/releases/1.8/#management-commands):
> Database connections are now always closed after a management command called from the command line has finished doing its job.

or maybe it's related to [Database connection reloading between tests](https://docs.djangoproject.com/en/1.9/releases/1.8/#database-connection-reloading-between-tests)

**Comment by nickpell** [comment]
@millerdev Looks like a prior test was not calling ```tearDownClass``` of ```TestCase```.  Thanks for suggesting to look at the previously run tests!

**Comment by nickpell** [comment]
@millerdev Any thoughts on this one?

```
======================================================================
ERROR: Failure: ImportError (No module named simple)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/vendor/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/vendor/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "submodules/dimagi-utils-src/dimagi/utils/couch/testrunner.py", line 4, in <module>
    from couchdbkit.ext.django.testrunner import CouchDbKitTestSuiteRunner
  File "/vendor/lib/python2.7/site-packages/couchdbkit/ext/django/testrunner.py", line 6, in <module>
    from django.test.simple import DjangoTestSuiteRunner
ImportError: No module named simple
```

**Comment by snopoke** [comment]
I suspect we can get rid of `testrunner.py` completely since we're now on
nose.

File: custom/ilsgateway/tests/handlers/messageinitatior.py [hunk 5]
```diff
@@ -1,237 +1,237 @@
...
-            32349 < %(response)s
-            """ % {
-            "test_handler_confirm": response1,
-            "response": response2
-        }
-        self.run_script(script)
+# from corehq.util.translation import localize
+# from custom.ilsgateway.models import SupplyPointStatus, SupplyPointStatusValues, SupplyPointStatusTypes
+# from custom.ilsgateway.tanzania.reminders import TEST_HANDLER_CONFIRM, SUBMITTED_REMINDER_FACILITY, \
+#     LOSS_ADJUST_HELP, TEST_HANDLER_BAD_CODE, DELIVERY_REMINDER_FACILITY, SOH_HELP_MESSAGE, SUPERVISION_REMINDER, \
```
Comment (dimagimon) [inline custom/ilsgateway/tests/handlers/messageinitatior.py]: E501 line too long (116 > 115 characters)

File: custom/ilsgateway/tests/test_report_runner.py [hunk 6]
```diff
@@ -17,185 +17,185 @@
...
+#             self.assertEqual(
+#                 GroupSummary.objects.filter(
+#                     org_summary__location_id=self.district2.get_id,
+#                     title=SupplyPointStatusTypes.SOH_FACILITY
+#                 ).first().total, 1
+#             )
+#
+#             self._move_location(self.facility1, self.district2.get_id)
+#
+#             pr = PendingReportingDataRecalculation.objects.filter(sql_location=self.facility1.sql_location).first()
```
Comment (dimagimon) [inline custom/ilsgateway/tests/test_report_runner.py]: E501 line too long (117 > 115 characters)

File: custom/ilsgateway/tests/test_report_runner.py [hunk 7]
```diff
@@ -17,185 +17,185 @@
...
+#
+#             self.assertEqual(
+#                 GroupSummary.objects.filter(
+#                     org_summary__location_id=self.facility2.get_id,
+#                     title=SupplyPointStatusTypes.R_AND_R_FACILITY
+#                 ).first().total, 1
+#             )
+#
+#             self._change_group(self.facility1, 'C')
+#             pr = PendingReportingDataRecalculation.objects.filter(sql_location=self.facility1.sql_location).first()
```
Comment (dimagimon) [inline custom/ilsgateway/tests/test_report_runner.py]: E501 line too long (117 > 115 characters)

File: requirements/requirements.txt [hunk 8]
```diff
@@ -7,7 +7,7 @@ jsonobject-couchdbkit==0.7.4.0
 celery==3.1.18
 # required by django-digest
 decorator==4.0.4
-django==1.7.11
+django==1.8.13
```
Comment (emord) [inline requirements/requirements.txt]: https://docs.djangoproject.com/en/1.10/releases/1.8.14/

**Comment by nickpell** [comment]
✅

**Pull request merged**